sim/src/bookinfo/reviews.rs

```diff
@@ -0,0 +1,202 @@
+                       &self.id, &self.capacity, &self.egress_rate, &self.generation_rate, &self.queue.size()),
+                       width = width)
+            } else {
+                write!(f, "{:width$}",
+                       &format!("Reviews {{ id : {}, capacity : {}, egress_rate : {}, generation_rate : {}, queue : {}, \n\tplugin : {} }}",
+                       &self.id, &self.capacity, &self.egress_rate, &self.generation_rate, &self.queue.size(), self.plugin.as_ref().unwrap()),
+                       width = width)
+            }
+        } else {
+            if self.plugin.is_none() {
+                write!(f, "Reviews {{ id : {}, egress_rate : {}, generation_rate : {}, plugin : None, capacity : {}, queue : {} }}",
+                       &self.id, &self.egress_rate, &self.generation_rate, &self.capacity, &self.queue.size())
+            } else {
+                write!(
+                    f,
+                    "Reviews {{ id : {}, egress_rate : {}, generation_rate : {}, plugin : {}, capacity : {}, queue : {} }}",
+                    &self.id,
+                    &self.egress_rate,
+                    &self.generation_rate,
+                    self.plugin.as_ref().unwrap(),
+                    &self.capacity,
+                    &self.queue.size()
+                )
+            }
+        }
+    }
+}
+
+impl SimElement for Reviews {
+    fn tick(&mut self, tick: u64) -> Vec<(Rpc, String)> {
+        let mut ret = vec![];
+        for _ in 0..min(
+            self.queue.size() + (self.generation_rate as usize),
+            self.egress_rate as usize,
+        ) {
+            let rpc: Rpc;
+            if self.queue.size() > 0 {
+                let deq = self.dequeue(tick);
+                rpc = deq.unwrap();
+            } else {
+                rpc = Rpc::new_rpc(&tick.to_string());
+            }
+            let neigh_len = self.neighbors.len();
+            if rpc.headers.contains_key("dest") {
+                let mut have_dest = false;
+                let dest = &rpc.headers["dest"].clone();
+                for n in &self.neighbors {
+                    if n == dest {
+                        have_dest = true;
+                        ret.push((rpc, dest.clone()));
+                        break;
+                    }
+                }
+                if !have_dest {
+                    print!("WARNING:  RPC given with invalid destination {0}\n", dest);
+                }
+            } else if neigh_len > 0 && rpc.headers["direction"] == "request" {
+                for neighbor in &self.neighbors {
+                    if neighbor.contains("ratings-v1") {
+                        ret.push((rpc, neighbor.to_string()));
+                        break;
+                    }
+                }
+            } else if neigh_len > 0 {
+                for neighbor in &self.neighbors {
+                    if neighbor.contains("productpage-v1") {
```

Comment:
Do not need this check. Reviews should probably just broadcast an incoming rpc to all other edges, both requests and responses. Since we only have two edges, that means they just get put out on the other edge.

Reply:
Makes sense.  Changed it so it just picks the other edge.